### ISSUE TITLE
Carousel wont rerender when slides array is empty

### ISSUE THREAD

**Issue by avrit** [vitalii-andriiovskyi/ngx-owl-carousel-o#242]
Ive implemented a filter to the slides and the filter founds no results, the new slides array should be empty.
When the new array of slides is empty the carousel wont rerender so although the new array of slides is empty the carousel show all the previous slides

**Comment by vitalii-andriiovskyi** [comment]
Hi @avrit 

How do you filter slides? I try to remove a slide in my testing application and it works, the slide gets removed. I remove all slides and leave empty array and the carousel shows no slides after some delay however. It works for me.

**Comment by avrit** [comment]
Im sending the carousel a new filtered array of slides via rxjs subject and the new array can be an empty array.
 
This is the code from carousel.component.ts:

ngOnChanges() {
    if (this.prevOptions !== this.options) {
      if (this.prevOptions && this.slides?.toArray().length) {
        this.carouselService.setup(this.carouselWindowWidth, this.slides.toArray(), this.options);
        this.carouselService.initialize(this.slides.toArray());
      } else if (this.prevOptions && !this.slides?.toArray().length) {
        this.carouselLoaded = false;
        this.logger.log(`There are no slides to show. So the carousel won't be re-rendered`);
      } else {
        this.carouselLoaded = false;
      }
      this.prevOptions = this.options;
    }
  }

As far as i understand this, when the array is empty, 'this.slides?.toArray().length will be equal to 0 and we will enter the second else if clause which will no re render the carousel. I get the message in the console that says that the carousel won't be re-rendered as it says in the second clause

**Comment by vitalii-andriiovskyi** [comment]
@avrit 

Everything in the `ngOnChanges` will work if you change your carousel options. 

```typescript
if (this.prevOptions !== this.options) {
   if (this.prevOptions && this.slides?.toArray().length) { 
     //...
    }
   // ....
}
```

But if you change only the array with slides data the options will remain the same and comparison `this.prevOptions !== this.options` will be `false` and nothing in this if clause will work. 

Please, share the code so I could be able to reproduce the problem.

**Comment by avrit** [comment]
Im not sure i understand. First i dont want to change the options i only want to change the slides array.
Second in the ngOnChanges even if the options has changes the array must not be empty in order to the carousel to re render.

also you can see it here:
ngAfterContentInit() {
    if (this.slides.toArray().length) {
      this.carouselService.setup(this.carouselWindowWidth, this.slides.toArray(), this.options);
      this.carouselService.initialize(this.slides.toArray());

      this._winResizeWatcher();
    } else {
      this.logger.log(`There are no slides to show. So the carousel won't be rendered`);
    }

    this._slidesChangesSubscription = this.slides.changes.pipe(
      tap((slides) => {
        if (slides.toArray().length) {
          // this.carouselService.setItems(slides.toArray());
          this.carouselService.setup(this.carouselWindowWidth, slides.toArray(), this.options);
          this.carouselService.initialize(slides.toArray());
        } else {
          this.carouselLoaded = false;
          this.logger.log(`There are no slides to show. So the carousel won't be re-rendered`);
        }
      })
    ).subscribe(()=>{});

  }

if the slides array is empty nothing happens

**Comment by vitalii-andriiovskyi** [comment]
@avrit 

Yes, you're right if the slides array is empty nothing happens. If it is empty at the beginning the carousel is not created at all. If it empty later it doesn't rerenders. Actually it should hide the carousel completely, cause there's nothing to show.

I think for now you can do this:
```HTML
<ng-container *ngIf="slidesData$ | async as slidesData; then carousel"></ng-container>

<ng-template #carousel>
  <ng-container *ngIf="slidesData?.length > 0">
    <owl-carousel-o>
        your slides
    </owl-carousel-o>
  </ng-container>
</ng-template>

```

**Comment by Arijit0** [comment]
> Ive implemented a filter to the slides and the filter founds no results, the new slides array should be empty. When the new array of slides is empty the carousel wont rerender so although the new array of slides is empty the carousel show all the previous slides

I think you forgot to put carouselSlide  inside ng-template. like this. 

`<ng-template  carouselSlide>
  <ng-container *ngIf="slidesData?.length > 0">
    <owl-carousel-o>
        your slides
    </owl-carousel-o>
  </ng-container>
</ng-template>`
